File: client/src/lib/types/advanced-analytics.ts
// Period Comparison Types
export interface PeriodComparisonItem {
  id: string;
  name: string;
  period_a_average: number | null;
  period_b_average: number | null;
  change: number | null;
  change_percent: number | null;
  student_count_a: number;
  student_count_b: number;
  teacher_name?: string;
  subject?: string;
  class_name?: string;
}

export interface PeriodComparisonResponse {
  comparison_type: "class" | "subject_teacher" | "subject";
  period_a: string;
  period_b: string;
  data: PeriodComparisonItem[];
}

// Red Student Types
export interface RedStudentGroup {
  id: string;
  name: string;
  red_student_count: number;
  total_student_count: number;
  percentage: number;
  average_grade: number;
}

export interface FailingSubject {
  subject: string;
  teacher_name: string | null;
  grade: number;
}

export interface RedStudentDetail {
  student_tz: string;
  student_name: string;
  class_name: string | null;
  grade_level: string | null;
  average_grade: number;
  failing_subjects: FailingSubject[];
}

export interface RedStudentSegmentation {
  total_red_students: number;
  threshold: number;
  by_class: RedStudentGroup[];
  by_layer: RedStudentGroup[];
  by_teacher: RedStudentGroup[];
  by_subject: RedStudentGroup[];
}

export interface RedStudentListResponse {
  total: number;
  page: number;
  page_size: number;
  students: RedStudentDetail[];
}

// Versus Comparison Types
export interface VersusSeriesItem {
  id: string;
  name: string;
  value: number;
  student_count: number;
  subjects?: string[];
  teacher_name?: string;
}

export interface VersusChartData {
  comparison_type: "class" | "teacher" | "layer";
  metric: string;
  series: VersusSeriesItem[];
}

// Cascading Filter Types
export interface ClassOption {
  id: string;
  class_name: string;
  grade_level: string;
}

export interface TeacherOption {
  id: string | null;
  name: string;
  subjects: string[];
}

export interface CascadingFilterOptions {
  classes: ClassOption[];
  teachers: TeacherOption[];
  subjects: string[];
}

File: server/src/schemas/advanced_analytics.py
"""Grades analytics schemas for period comparison and risk analysis."""

from pydantic import BaseModel


class PeriodComparisonItem(BaseModel):
    """Single item in period comparison."""

    id: str
    name: str
    period_a_average: float | None
    period_b_average: float | None
    change: float | None
    change_percent: float | None
    student_count_a: int
    student_count_b: int
    teacher_name: str | None = None
    subject: str | None = None
    class_name: str | None = None


class PeriodComparisonResponse(BaseModel):
    """Response for period comparison endpoint."""

    comparison_type: str
    period_a: str
    period_b: str
    data: list[PeriodComparisonItem]


class RedStudentGroup(BaseModel):
    """Grouping of at-risk students."""

    id: str
    name: str
    red_student_count: int
    total_student_count: int
    percentage: float
    average_grade: float


class FailingSubject(BaseModel):
    """Subject where student is failing."""

    subject: str
    teacher_name: str | None
    grade: float


class RedStudentDetail(BaseModel):
    """Detailed at-risk student info."""

    student_tz: str
    student_name: str
    class_name: str | None
    grade_level: str | None
    average_grade: float
    failing_subjects: list[FailingSubject]


class RedStudentSegmentation(BaseModel):
    """Full red student segmentation response."""

    total_red_students: int
    threshold: float
    by_class: list[RedStudentGroup]
    by_layer: list[RedStudentGroup]
    by_teacher: list[RedStudentGroup]
    by_subject: list[RedStudentGroup]


class RedStudentListResponse(BaseModel):
    """Paginated list of red students."""

    total: int
    page: int
    page_size: int
    students: list[RedStudentDetail]


class VersusSeriesItem(BaseModel):
    """Single series in versus comparison."""

    id: str
    name: str
    value: float
    student_count: int
    subjects: list[str] = []
    teacher_name: str | None = None


class VersusChartData(BaseModel):
    """Response for versus comparison chart."""

    comparison_type: str
    metric: str
    series: list[VersusSeriesItem]


class ClassOption(BaseModel):
    """Class option for cascading filters."""

    id: str
    class_name: str
    grade_level: str


class TeacherOption(BaseModel):
    """Teacher option for cascading filters."""

    id: str | None
    name: str
    subjects: list[str]


class CascadingFilterOptions(BaseModel):
    """Available options for cascading filters."""

    classes: list[ClassOption]
    teachers: list[TeacherOption]
    subjects: list[str]
